packages/esbuild-plugin/src/index.js

```diff
@@ -0,0 +1,176 @@
+
+        const collectedCSS = stylexBabelPlugin.processStylexRules(
+          rules,
+          useCSSLayers,
+        );
+        const shouldWriteToDisk =
+          build.initialOptions.write === undefined ||
+          build.initialOptions.write;
+
+        if (shouldWriteToDisk) {
+          await mkdir(path.dirname(generatedCSSFileName), {
+            recursive: true,
+          });
+          await writeFile(generatedCSSFileName, collectedCSS, 'utf8');
+
+          return;
+        }
+
+        if (outputFiles !== undefined) {
+          outputFiles.push({
+            path: '<stdout>',
+            contents: new TextEncoder().encode(collectedCSS),
+            hash: utils.hash(new Date().getTime().toString()),
+            get text() {
+              return collectedCSS;
+            },
+          });
+        }
+      });
+
+      build.onLoad(
+        { filter: STYLEX_PLUGIN_ONLOAD_FILTER },
+        async (args: OnLoadArgs) => {
+          const currFilePath = args.path;
+          const inputCode = await readFile(currFilePath, 'utf8');
+
+          if (
+            !stylexImports.some((importName) => inputCode.includes(importName))
+          ) {
+            // avoid transform if file doesn't have stylex imports
+            // esbuild proceeds to the next callback
+            return;
+          }
+
+          const transformResult = await transformAsync(inputCode, {
+            babelrc: false,
+            filename: currFilePath,
+            presets,
+            plugins: [
+              ...plugins,
+              ...getFlowOrTypeScriptBabelSyntaxPlugins(currFilePath),
+              jsxSyntaxPlugin,
+              stylexBabelPlugin.withOptions({
+                ...options,
+                dev,
+                unstable_moduleResolution,
+              }),
+            ],
+          });
+
+          const loader = getEsbuildLoader(currFilePath);
+
+          if (transformResult === null) {
+            console.warn('StyleX: transformAsync returned null');
+            return { code: inputCode, loader };
+          }
+
+          const { code, metadata } = transformResult;
+
+          if (code === null) {
+            console.warn('StyleX: transformAsync returned null code');
+            return { code: inputCode, loader };
+          }
+
+          if (
+            !dev &&
+            (metadata: $FlowFixMe).stylex !== null &&
+            (metadata: $FlowFixMe).stylex.length > 0
+          ) {
+            stylexRules[args.path] = (metadata: $FlowFixMe).stylex;
+          }
+
+          return {
+            contents: code,
+            loader,
+          };
+        },
+      );
+    },
+  };
+}
+
+function getEsbuildLoader(fileName: string) {
+  if (fileName.endsWith('.tsx')) {
+    return 'tsx';
+  }
+
+  if (fileName.endsWith('.jsx')) {
+    return 'jsx';
+  }
+
+  if (fileName.endsWith('ts')) {
+    return 'ts';
+  }
+
+  return 'js';
+}
+
+function getFlowOrTypeScriptBabelSyntaxPlugins(fileName: string) {
```

Comment:
We also need to add check to verify that we only attempt to transform JS files. This means ignoring `.json` files.

This regex might help.
```js
/[mc]?[jt]sx?$/
```

Reply:
the context in which this function is used guarantees that we only transform JS files:

https://github.com/facebook/stylex/blob/cd1f8b94a1856bad591d34dcfdd4e325416fa016/packages/esbuild-plugin/src/index.js#L93

I assume we don't want to allow processing invalid filetypes like `.mjsx` or `.cjsx`, hence I added an explicit filter above:

https://github.com/facebook/stylex/blob/cd1f8b94a1856bad591d34dcfdd4e325416fa016/packages/esbuild-plugin/src/index.js#L28